### DISCUSSION TITLE
How to save logs to a file in Prisma

### DISCUSSION THREAD

**Discussion by 1tang2bang92** [prisma/prisma#24841] [Q&A]
### Question

Hello,
I'm currently working on a project using Prisma. I'd like to save the logs related to database queries to a file. This would help me monitor the application's performance and assist in troubleshooting.

Specifically, I'd like to know:
1. Is there a default setting in Prisma to save logs to a file?
2. If not, how can I implement this functionality?
3. Can I set log levels to save only specific types of queries?

Any advice or example code would be greatly appreciated. Thank you.
Is this translation suitable for your needs? Would you like me to modify or expand on anything?

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hello @1tang2bang92 👋 

Thank you for raising this queston.

Prisma does not provide a default setting to save logs directly to a file. However, you can implement this functionality using event-based logging. Here are the steps and example code to help you achieve this:

1. Set up event-based logging: Configure Prisma Client to emit log events.
2. . Subscribe to log events: Use the `$on()` method to handle log events.
3. Write logs to a file: Use Node.js's fs module to write the logs to a file.

```js
const { PrismaClient, Prisma, Result  } = require("./generated/client")
const fs = require('fs');
const path = require('path');

// Create a write stream (in append mode)
const logFile = fs.createWriteStream(path.join(__dirname, 'prisma-queries.log'), { flags: 'a' });

const prisma = new PrismaClient({
  log: [
    {
      emit: 'event',
      level: 'query',
    },
    {
      emit: 'stdout',
      level: 'error',
    },
    {
      emit: 'stdout',
      level: 'info',
    },
    {
      emit: 'stdout',
      level: 'warn',
    },
  ],
});

async function main() {

  prisma.$on('query', (e) => {
    const logMessage = `Query: ${e.query}\nParams: ${e.params}\nDuration: ${e.duration}ms\n\n`;
    logFile.write(logMessage);
    console.log(logMessage); // Optional: also log to console
  });
const post = await prisma.post.create({ 
  data: { 

    title: 'My Awesome Post!', 

    content: 'This is the content of my post.', 

  }, 

})

 console.log(`Created post with ID: ${post.id}`); 

     
};


main()
  .catch(e => {
    throw e
  })
  .finally(async () => {
    await prisma.$disconnect()
  })

```
You can configure Prisma Client to log specific types of queries by setting the appropriate log levels. In the example above, the log levels are set to query, error, info, and warn. You can adjust these levels based on your needs.

**Reply by 1tang2bang92** [reply]
answer is very helpful. Thank you!
